protocol/l1-upgrades.md

```diff
@@ -0,0 +1,329 @@
+`ProxyAdmin`), since it is known already for each contract, ie. the only non-standard Proxies are
+the `L1StandardBridge` (`L1ChugsplashProxy`) and `L1CrossDomainMessenger` (`ResolvedDelegateProxy`).
+
+## Contract versioning
+
+As part of this work, we will cease to version each L1 contract, and instead move to a global L1
+contracts version. This version will be stored in the `SystemConfig` and OPCM. For backwards
+compatibility, the existing `version()` getters will return the value from `SystemConfig.version()`.
+
+The `DeployImplementations` script will write the version into the OPCM, which writes it into the
+`SystemConfig`.
+
+The `rc` tag will be included on the OPCM's initial version. It will be programmatically removed
+when `upgrade()` is called by the Upgrade Controller multisig, which signifies governance approval.
+
+## Proof system considerations
+
+### Adding the PermissionlessDisputeGame to a chain
+
+Given that not all chains will support the `PermissionlessDisputeGame` upon deployment, an
+`OPCM.addGameType()` method will be added which will handle this.
+
+WIP
+
+1. ideally we update op-program to remove the chain config, so that it is identical across chains
+   then it reads the config at run time.
+   But we won't block isthmus on it, and will allow the `FaultDisputeGame` and `PermissionedDisputeGame` to be non-compliant.
+2. inputs for each hardfork: one prestatehash per chain.
+
+## Release process
+
+As part of the release process, the associated implementation contracts and a new OPCM, will be deployed.
+
+The Upgrade Controller (which MUST be a Safe), will `DELEGATECALL` the `OPCM.upgrade()` function,
+providing the address of the `SuperchainProxyAdmin` and a list of the `SystemConfig` contracts for the
+OP Chains to be upgraded.
+
+Importantly,
+
+Thus the high level logic for upgrading a contract should be roughly as follows:
+
+```js
+
+// Some upgrades will
+struct NewChainConfig {
+  address newValues;
+}
+
+Addresses immutable implementations;
+
+constructor(Addresses _implementations) {
+  implementations = _implementations;
+}
+
+function upgrade(SuperchainProxyAdmin _admin, ISystemConfig[] _systemConfigs, NewChainConfig[] _newConfigs) public {
+  for(uint i=0; i< systemConfigs.length; i++) {
+    // Read the `Addresses` struct from each `SystemConfig` in the `systemConfigs` mapping.
+    // For each entry in the `Addresses` struct:
+      // 1. Call the `SuperchainProxyAdmin` to update the implementation to the `Populator` contract
+      //    TODO: OR update the implementation to
+      // 2. call
+      // 2. Update the implementation to the final implementation contract.
+  }
+  // Return ownership of the `SuperchainProxyAdmin` to the Upgrade Controller (`msg.sender`)
+      //    `SuperchainProxyAdmin`. The unique upgrade call encoding required for the non-standard proxies
+      //    should be handled here.
+}
+```
```

Comment:
Am I correct in understanding that the upgrade function called with a DELEGATECALL in the safe is mostly just passed a list of addresses for the currently deployed contracts & the upgrad populator contracts? And all these contracts / all the upgrade function calldata is stored somewhere else?

Assuming my understanding is roughly correct, my main concern with this approach is that it seems extremely stateful. I really don't want there to be a footgun where we can deploy populator contracts or set upgrade values and then somehow those contracts change by the time we call upgrade and there's a diff in how the upgrade is executed.

It would be nice if the `upgrade(..)` function that the safe DELEGATECALLs to included as input data all information required for the upgrade. That way we just have one upgrade payload that we need to audit that contains all information utilized in the upgrade. I just don't want to have to look in a bunch of different places to find all the info I need to audit an upgrade if that makes sense

Reply:
Good callout, I share your concern.


> It would be nice if the upgrade(..) function that the safe DELEGATECALLs to included as input data all information required for the upgrade. 

Agree. What we can do is use the `pure` modifier on `upgrade()` to ensure that no reads or writes occur in the Safe.

Reply:
Overall flow I am envisioning based on the most recent conversation I had with @maurelian:
- Deploy new `OPCM` contract, which contains:
  - `deploy` and `upgrade` methods to deploy new chains and ugprade existing chains
  - It also contains immutables pointing to all new implementation contracts (e.g. a new SystemConfig contract)
  - For each new implementation it knows how to to encode it's e.g. `SystemConfig.upgrade` call, and any chain-specific arguments to that call are passed into the root OPCM `upgrade` method
  - Sample `OPCM.upgrade` sig is `upgrade(systemConfig, arg1, arg2)`
- A safe delegate calls to the `OPCM.upgrade(...)` method
- The upgrade flow that OPCM does for each new implementation contract is a two step flow where first we set the implementation to a contract like this:
```solidity
contract UpgradePreparer {
  // Same initializer slot used by OpenZeppelin v5
  // keccak256(abi.encode(uint256(keccak256("openzeppelin.storage.Initializable")) - 1)) & ~bytes32(uint256(0xff))
  bytes32 private constant INITIALIZABLE_STORAGE = 0xf0c57e16840df040f15088dc2f81fe391c3923bec73e23a9662efc9c229c6a00;
  function prepare() external {

    // Clear the initializer slot so we can call `upgrade` on the new implementation
    assembly {
      sstore(INITIALIZABLE_STORAGE, 0)
    }
  }
}
```
- We call `prepare` to clear the slot
- Then we put the new `SystemConfig` contract in place and call it's `upgrade()` method. This sets the `INITIALIZABLE_STORAGE` slot to 1, and runs all applicable logic
- I am in favor of https://github.com/ethereum-optimism/design-docs/pull/159#discussion_r1852291740 which suggests always doing this two step upgrade flow even when `upgrade` is a no-op, as it simplifies the number of paths we have to worry about and reduces the chance of us forgetting an `upgrade` call

Reply:
I've ensured that the flow above is outlined in sufficient detail in the contents of the doc. 

Keeping this open as a place to discuss the use of DELEGATECALL in particular.